<a id="section-one"></a>
**<font color = DarkBlue size = 5>Import related libraries</font>**

First of all we have to import all the needed libraries:


- Numpy: a library used for working with arrays. It also has functions for working in domain of linear algebra, fourier transform, and matrices.

- Pandas: a library offers data structures and operations for manipulating numerical tables and time series.

- Pandas_Profiling: an open source Python library with which we can quickly do an exploratory data analysis with just a few lines of code.

- Matplotlib: a plotting library for the Python programming language and its numerical mathematics extension NumPy

- Sklearn:Scikit-learn is a free software machine learning library for the Python programming language. 


In [2]:
# import libraries
import numpy as np
import pandas as pd 
from pandas_profiling import ProfileReport
import matplotlib.pyplot as plt
from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

<a id="section-two"></a>
**<font color = DarkBlue size = 5>Read data and Inspect Data</font>**

In [3]:
#load dataset
df = pd.read_csv("owid-covid-data.csv")

In [4]:
#Show first five rows of data
df.head()

iso_code continent     location        date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan  2020-02-24          5.0        5.0   
1      AFG      Asia  Afghanistan  2020-02-25          5.0        0.0   
2      AFG      Asia  Afghanistan  2020-02-26          5.0        0.0   
3      AFG      Asia  Afghanistan  2020-02-27          5.0        0.0   
4      AFG      Asia  Afghanistan  2020-02-28          5.0        0.0   

   new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
0                 NaN           NaN         NaN                  NaN  ...   
1                 NaN           NaN         NaN                  NaN  ...   
2                 NaN           NaN         NaN                  NaN  ...   
3                 NaN           NaN         NaN                  NaN  ...   
4                 NaN           NaN         NaN                  NaN  ...   

   female_smokers  male_smokers  handwashing_facilities  \
0             NaN           NaN                  37.746   
1             NaN           NaN                  37.746   
2             NaN           NaN                  37.746   
3             NaN           NaN                  37.746   
4             NaN           NaN                  37.746   

   hospital_beds_per_thousand  life_expectancy  human_development_index  \
0                         0.5            64.83                    0.511   
1                         0.5            64.83                    0.511   
2                         0.5            64.83                    0.511   
3                         0.5            64.83                    0.511   
4                         0.5            64.83                    0.511   

   excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
0                                   NaN                          NaN   
1                                   NaN                          NaN   
2                                   NaN                          NaN   
3                                   NaN                          NaN   
4                                   NaN                          NaN   

   excess_mortality  excess_mortality_cumulative_per_million  
0               NaN                                      NaN  
1               NaN                                      NaN  
2               NaN                                      NaN  
3               NaN                                      NaN  
4               NaN                                      NaN  

[5 rows x 65 columns]

In [5]:
# Information of data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132644 entries, 0 to 132643
Data columns (total 65 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   iso_code                                 132644 non-null  object 
 1   continent                                124170 non-null  object 
 2   location                                 132644 non-null  object 
 3   date                                     132644 non-null  object 
 4   total_cases                              125410 non-null  float64
 5   new_cases                                125408 non-null  float64
 6   new_cases_smoothed                       124365 non-null  float64
 7   total_deaths                             114293 non-null  float64
 8   new_deaths                               114489 non-null  float64
 9   new_deaths_smoothed                      124365 non-null  float64
 10  total_cases_per_million         

In [6]:
# Pandas Profiling
rep = ProfileReport(df)
rep

In [7]:
# View column data types
df.dtypes

iso_code                                    object
continent                                   object
location                                    object
date                                        object
total_cases                                float64
                                            ...   
human_development_index                    float64
excess_mortality_cumulative_absolute       float64
excess_mortality_cumulative                float64
excess_mortality                           float64
excess_mortality_cumulative_per_million    float64
Length: 65, dtype: object

Result:
- There are 65 columns of object and float datatypes

In [8]:
# Check Duplicates
df.duplicated().sum()

0

Result:
- Dataset don't contains any duplicate rows.


In [9]:
# Get statistical data about each column
df.describe(include="all") 

iso_code continent location        date   total_cases      new_cases  \
count    132644    124170   132644      132644  1.254100e+05  125408.000000   
unique      237         6      237         681           NaN            NaN   
top         ARG    Africa     Peru  2021-07-14           NaN            NaN   
freq        681     32809      681         234           NaN            NaN   
mean        NaN       NaN      NaN         NaN  2.038985e+06    8353.112888   
std         NaN       NaN      NaN         NaN  1.163910e+07   43547.802492   
min         NaN       NaN      NaN         NaN  1.000000e+00  -74347.000000   
25%         NaN       NaN      NaN         NaN  2.388000e+03       3.000000   
50%         NaN       NaN      NaN         NaN  2.714850e+04     104.000000   
75%         NaN       NaN      NaN         NaN  2.641010e+05    1077.000000   
max         NaN       NaN      NaN         NaN  2.519418e+08  907963.000000   

        new_cases_smoothed  total_deaths     new_deaths  new_deaths_smoothed  \
count        124365.000000  1.142930e+05  114489.000000        124365.000000   
unique                 NaN           NaN            NaN                  NaN   
top                    NaN           NaN            NaN                  NaN   
freq                   NaN           NaN            NaN                  NaN   
mean           8370.887531  5.045250e+04     184.004114           168.641989   
std           43035.659063  2.589636e+05     871.591559           817.796127   
min           -6223.000000  1.000000e+00   -1918.000000          -232.143000   
25%              10.286000  8.000000e+01       0.000000             0.143000   
50%             130.143000  7.410000e+02       2.000000             2.000000   
75%            1133.429000  6.468000e+03      22.000000            18.571000   
max          826457.571000  5.079990e+06   18007.000000         14703.286000   

        ...  female_smokers  male_smokers  handwashing_facilities  \
count   ...    89329.000000  88037.000000            57713.000000   
unique  ...             NaN           NaN                     NaN   
top     ...             NaN           NaN                     NaN   
freq    ...             NaN           NaN                     NaN   
mean    ...       10.589701     32.748528               50.865182   
std     ...       10.502683     13.512865               31.821458   
min     ...        0.100000      7.700000                1.188000   
25%     ...        1.900000     21.600000               20.859000   
50%     ...        6.300000     31.400000               49.839000   
75%     ...       19.300000     41.300000               83.241000   
max     ...       44.000000     78.100000              100.000000   

        hospital_beds_per_thousand  life_expectancy  human_development_index  \
count                104208.000000    123313.000000            114771.000000   
unique                         NaN              NaN                      NaN   
top                            NaN              NaN                      NaN   
freq                           NaN              NaN                      NaN   
mean                      3.029443        73.260475                 0.726207   
std                       2.455593         7.532205                 0.150042   
min                       0.100000        53.280000                 0.394000   
25%                       1.300000        67.920000                 0.602000   
50%                       2.400000        74.620000                 0.744000   
75%                       4.000000        78.740000                 0.845000   
max                      13.800000        86.750000                 0.957000   

        excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
count                            4656.000000                  4656.000000   
unique                                   NaN                          NaN   
top                                      NaN                          NaN   

Result:
- Dataset contains 65 columns and almost each column conatins NaN


In [10]:
# Draw Histogram to show distribution of data for each feature
df.hist(figsize=(35,35))
plt.suptitle('Histograms', fontsize=18);
plt.savefig("Histogram.png")

Result:
- A Histogram of all the 65 columns


In [12]:
#copy dataset into another variable
df_original=df

In [13]:
#copy dataset bcz sometime we have to again start from this point 
df=df_original

In [11]:
df.dtypes

iso_code                                    object
continent                                   object
location                                    object
date                                        object
total_cases                                float64
                                            ...   
human_development_index                    float64
excess_mortality_cumulative_absolute       float64
excess_mortality_cumulative                float64
excess_mortality                           float64
excess_mortality_cumulative_per_million    float64
Length: 65, dtype: object

Result:
- Dataset have different data types


In [12]:
# we don't need this column because we will follow number of rows for time and date
df=df.drop('date', axis=1)
df.shape

(132644, 64)

Result:
- After droping the column 'date' we have 132644 rows and 64 columns 


In [13]:
# remove this colum also
df=df.drop('location', axis=1)
df.shape

(132644, 63)

Result:
- After droping the column 'location' we have 132644 rows and 63 columns 


In [14]:
# remove this colum also
df=df.drop('continent', axis=1)
df.shape

(132644, 62)

Result:
- After droping the column 'continent' we have 132644 rows and 62 columns 


In [15]:
# remove this colum also
df=df.drop('iso_code', axis=1)
df.shape

(132644, 61)

Result:
- After droping the column 'iso-code' we have 132644 rows and 61 columns 


In [16]:
# remove this colum also
df=df.drop('tests_units', axis=1)
df.shape

(132644, 60)

Result:
- After droping the column 'tests_units' we have 132644 rows and 60 columns 


In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132644 entries, 0 to 132643
Data columns (total 60 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   total_cases                              125410 non-null  float64
 1   new_cases                                125408 non-null  float64
 2   new_cases_smoothed                       124365 non-null  float64
 3   total_deaths                             114293 non-null  float64
 4   new_deaths                               114489 non-null  float64
 5   new_deaths_smoothed                      124365 non-null  float64
 6   total_cases_per_million                  124766 non-null  float64
 7   new_cases_per_million                    124764 non-null  float64
 8   new_cases_smoothed_per_million           123726 non-null  float64
 9   total_deaths_per_million                 113662 non-null  float64
 10  new_deaths_per_million          

In [18]:
df.dtypes

total_cases                                float64
new_cases                                  float64
new_cases_smoothed                         float64
total_deaths                               float64
new_deaths                                 float64
new_deaths_smoothed                        float64
total_cases_per_million                    float64
new_cases_per_million                      float64
new_cases_smoothed_per_million             float64
total_deaths_per_million                   float64
new_deaths_per_million                     float64
new_deaths_smoothed_per_million            float64
reproduction_rate                          float64
icu_patients                               float64
icu_patients_per_million                   float64
hosp_patients                              float64
hosp_patients_per_million                  float64
weekly_icu_admissions                      float64
weekly_icu_admissions_per_million          float64
weekly_hosp_admissions         

In [19]:
#count NaN 
df.isna().sum()

total_cases                                  7234
new_cases                                    7236
new_cases_smoothed                           8279
total_deaths                                18351
new_deaths                                  18155
new_deaths_smoothed                          8279
total_cases_per_million                      7878
new_cases_per_million                        7880
new_cases_smoothed_per_million               8918
total_deaths_per_million                    18982
new_deaths_per_million                      18786
new_deaths_smoothed_per_million              8918
reproduction_rate                           28218
icu_patients                               116697
icu_patients_per_million                   116697
hosp_patients                              113952
hosp_patients_per_million                  113952
weekly_icu_admissions                      131304
weekly_icu_admissions_per_million          131304
weekly_hosp_admissions                     130518


Result:
- Each column contain NaN values 


In [20]:
# replace NaN with 0
df =df.fillna(0)


Result:
- Replace NaN with zero becaue there should not be NaN values 


In [21]:
#count NaN 
df.isna().sum()

total_cases                                0
new_cases                                  0
new_cases_smoothed                         0
total_deaths                               0
new_deaths                                 0
new_deaths_smoothed                        0
total_cases_per_million                    0
new_cases_per_million                      0
new_cases_smoothed_per_million             0
total_deaths_per_million                   0
new_deaths_per_million                     0
new_deaths_smoothed_per_million            0
reproduction_rate                          0
icu_patients                               0
icu_patients_per_million                   0
hosp_patients                              0
hosp_patients_per_million                  0
weekly_icu_admissions                      0
weekly_icu_admissions_per_million          0
weekly_hosp_admissions                     0
weekly_hosp_admissions_per_million         0
new_tests                                  0
total_test

Result:
- After replacing NaN, there is no NaN values


In [22]:
# checking the min and max values
df['new_deaths'].max()

18007.0

Result:
- In a single day, 18007 is number maxium of deaths


In [23]:
# Max value at index
df['new_deaths'].idxmax()

130561

Result:
- Number of row in dataset which contains 18007 death per day 


In [24]:
# min value
df['new_deaths'].min()

-1918.0

Result:
- In a single day, -1918 is minimum number of deaths but it is wrong. Because minimum death should be zero, not less than zero


In [25]:
# min value at index
df['new_deaths'].idxmin()

111824

Result:
- Number of row in dataset which contains -1918 death per day 

In [26]:
# total number of negative values
sum(n < 0 for n in df['new_deaths'].values.flatten())

124

Result:
- Total number of negative values in 'new_deaths'

In [27]:
# replace negative values to 0
df['new_deaths']=df['new_deaths'].clip(lower=0)

Result:
- Replace the all the negative values in 'new_deaths' with 0

In [28]:
# again check total number of negative values
sum(n < 0 for n in df['new_deaths'].values.flatten())

0

Result:
- Total number of negative values in 'new_deaths' after replacing with 0

In [29]:
# convert complete dataset to float
df =df.astype(float)

Result:
- Convert dataset to float data type for further used by Algorithms

In [30]:
df.dtypes

total_cases                                float64
new_cases                                  float64
new_cases_smoothed                         float64
total_deaths                               float64
new_deaths                                 float64
new_deaths_smoothed                        float64
total_cases_per_million                    float64
new_cases_per_million                      float64
new_cases_smoothed_per_million             float64
total_deaths_per_million                   float64
new_deaths_per_million                     float64
new_deaths_smoothed_per_million            float64
reproduction_rate                          float64
icu_patients                               float64
icu_patients_per_million                   float64
hosp_patients                              float64
hosp_patients_per_million                  float64
weekly_icu_admissions                      float64
weekly_icu_admissions_per_million          float64
weekly_hosp_admissions         

In [31]:
# handle NaN and infinite values
df = df.reset_index()

Result:
- Handle NaN value if there occur in dataset

In [32]:
# here save a column as our label or target-column
target_col=np.array(df['new_deaths'])

Result:
- Set 'new_deaths' column as our target column because we are going to train an algorithm to predict the number new death which may be in near future.

In [33]:
# remove column
df=df.drop('new_deaths', axis=1)
df.shape

(132644, 60)

Result:
- Becaue this column is target column therefore we keep it 'Y' and other dataset to 'X'

In [34]:
df = df.reset_index()

In [35]:
#copy dataset for forther use
mydataset2 = df;
mydataset2.shape

(132644, 61)

In [36]:
#saving dataset columns as list because we will use it for feature selection process
mydataset_list=list(df.columns)

In [37]:
#convert dataset into array
mydataset=np.array(df)

In [38]:
#copy dataset which is in watt and array form
copy_mydataset=mydataset;


In [39]:
# split to 80% for training and 20% for testing 
mydataset_train, mydataset_test, target_col_train, target_col_test=train_test_split(mydataset,target_col, test_size=0.20, random_state=7)

Result:
- Split the dataset into train and test. 

In [40]:
#print into of spliting data
print('Training Features Shape:', mydataset_train.shape)
print('Training Labels Shape:', target_col_train.shape)
print('Testing Features Shape:', mydataset_test.shape)
print('Testing Labels Shape:', target_col_test.shape)
print('Dataset Shape:', mydataset.shape)

Training Features Shape: (106115, 61)
Training Labels Shape: (106115,)
Testing Features Shape: (26529, 61)
Testing Labels Shape: (26529,)
Dataset Shape: (132644, 61)


In [ ]:
#............Fit Classifier....#
nNNAC = KNeighborsClassifier(n_neighbors=20) 
nNNAC.fit(mydataset_train, target_col_train)


Result:
- Apply K Nearest Neighbor classifier with 'n_neighbors=20', we can update this hyperparameter value.
- Train with training sample

In [ ]:
# predict at test data
predictions1 = nNNAC.predict(mydataset_test)

Result:
- Input unseen data to algorithm

In [ ]:
# Plot of load forcasted by K-NN Classifier    
plt.figure()
plt.ylabel('Death');
plt.xlabel('Days'); 
plt.plot(target_col_test,'-g',label='Actual'); 
plt.plot(predictions1,'-r',label='KNN')
plt.title('Death_K-NN')
plt.gca().set_xlim(left=1); plt.legend(fancybox=True, framealpha=0.5)
plt.gca().legend(('Actual','KNN'));
plt.savefig('DeathForecastingKNN.png',bbox_inches='tight',transparent='true')
plt.show();

In [ ]:
# evaluation matrices
print(metrics.accuracy_score(target_col_test, predictions1))

0.4849787025519243


Result:
- Predicted Accuracy
- Because dataset is very large and contains numerious number of eatures, therefore algorithm show low accuracy


<a id="subsection-three"></a>
**<font color = Teal size = 4 >Summary</font>**

**We lead dataset, analyse, generate a consize report, clean, pre-process, split into 80% and 20% ratio for algorithm training and testing, train K Nearest Neighbor algorithm, and evaluate the algorithm. It can also predict the number of upcomming new cases. This is beneficial for governemtn to predict the new cases and death and take appropriate action in order to overcome that figure**